### ISSUE TITLE
[varLib] Merging vpal GPOS feature can't handle differing number of subtables

### ISSUE THREAD

**Issue by readroberts** [fonttools/fonttools#1624]
The vpal GPOS feature can be constructed differently in different source fonts. A glyph may require both and xAdvance and xPlacement adjustment in one font, and only a xAdvance in another fonts. Also in one font, a long sequence of glyphs may use the same value record, but the same sequence of glyphs may use the same value record format, but have different adjustment values. These differences lead to the source fonts having a different number of subtables after optimization for size. The current code can't merge the features if the source fonts differ in number of subtables, or if the subtables  differ in coverage or number of value records. I attached a small test case. The GPOS data is edited in rather arbitrarily. If you have the afdko, you can build this with 'buildcff2vf -d test_vpal.designspace'.
[test_vpal.zip](https://github.com/fonttools/fonttools/files/3230051/test_vpal.zip)

**Comment by miguelsousa** [comment]
The test files provided can be exercised with this command,

    fonttools varLib --master-finder master_vpal_test/{stem}.otf test_vpal.designspace

This is the traceback it will produce:

```
Merging OpenType Layout tables
Traceback (most recent call last):
  File "fonttools", line 10, in <module>
    sys.exit(main())
  File "fontTools/__main__.py", line 29, in main
    runpy.run_module(mod, run_name='__main__')
  File "lib/python3.7/runpy.py", line 208, in run_module
    return _run_code(code, {}, init_globals, run_name, mod_spec)
  File "lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "fontTools/varLib/__main__.py", line 7, in <module>
    sys.exit(main())
  File "fontTools/varLib/__init__.py", line 1064, in main
    optimize=options.optimize
  File "fontTools/varLib/__init__.py", line 916, in build
    _merge_OTL(vf, model, master_fonts, axisTags)
  File "fontTools/varLib/__init__.py", line 688, in _merge_OTL
    merger.mergeTables(font, master_fonts, ['GSUB', 'GDEF', 'GPOS'])
  File "fontTools/varLib/merger.py", line 108, in mergeTables
    for m in master_ttfs])
  File "fontTools/varLib/merger.py", line 950, in mergeThings
    super(VariationMerger, self).mergeThings(out, lst)
  File "fontTools/varLib/merger.py", line 94, in mergeThings
    self.mergeObjects(out, lst)
  File "fontTools/varLib/merger.py", line 73, in mergeObjects
    mergerFunc(self, value, values)
  File "fontTools/varLib/merger.py", line 950, in mergeThings
    super(VariationMerger, self).mergeThings(out, lst)
  File "fontTools/varLib/merger.py", line 94, in mergeThings
    self.mergeObjects(out, lst)
  File "fontTools/varLib/merger.py", line 73, in mergeObjects
    mergerFunc(self, value, values)
  File "fontTools/varLib/merger.py", line 950, in mergeThings
    super(VariationMerger, self).mergeThings(out, lst)
  File "fontTools/varLib/merger.py", line 94, in mergeThings
    self.mergeObjects(out, lst)
  File "fontTools/varLib/merger.py", line 73, in mergeObjects
    mergerFunc(self, value, values)
  File "fontTools/varLib/merger.py", line 950, in mergeThings
    super(VariationMerger, self).mergeThings(out, lst)
  File "fontTools/varLib/merger.py", line 96, in mergeThings
    self.mergeLists(out, lst)
  File "fontTools/varLib/merger.py", line 82, in mergeLists
    self.mergeThings(value, values)
  File "fontTools/varLib/merger.py", line 950, in mergeThings
    super(VariationMerger, self).mergeThings(out, lst)
  File "fontTools/varLib/merger.py", line 92, in mergeThings
    mergerFunc(self, out, lst)
  File "fontTools/varLib/merger.py", line 755, in merge
    merger.mergeLists(self.SubTable, subtables)
  File "fontTools/varLib/merger.py", line 79, in mergeLists
    assert allEqualTo(out, lst, len), (len(out), [len(v) for v in lst])
AssertionError: ((3, [3, 2]), 'Lookup', '[0]', 'list', '.Lookup', 'LookupList', '.LookupList', 'GPOS', '.table', 'table_G_P_O_S_')
```

**Comment by readroberts** [comment]
@anthrotype Looks to me as if this is pretty easy to fix. 
In the merge function at  varLib.merger.py::line 730, there is already logic to canonicalize PairPos2 lookups, and clean up after the merge. 

For Single Pos lookups, I would first force the lookup to be mergeable:
- look through all the subtables to determine the minimum format value than can hold all of the value records used in any subtable
- flatten the subtables to a single subtable, and set all record formats to this format value.

The lookup can then be merged.

I would then optimize the lookup by the usual methods:
-sort single pos records by first minimum record type, then by record value, then by GID.
-build SinglePos format 1, with a new subtable for each change of record type and value
-build Single Pos format 2, with a new subtable for each change of record type
-use whichever was smaller.

Does this sound reasonable?

**Comment by anthrotype** [comment]
Thanks. That sounds reasonable to me, although @behdad is more familiar with that part of the code and may like to comment as well.

**Comment by behdad** [comment]
That is the correct approach in general.  Regarding the optimization scheme I'm not sure about the details.  I have not been able to devise a satisfactory scheme for how to break down SinglePos into subtables..  You should, however, just use `otlLib.builder.buildSinglePos`.  We can separately discuss how to improve that function.  I think there's a bug open about otlLib.builder.buildSinglePos creating too many subtables..  I'll comment separately about possible improvements to that function.

> -sort single pos records by first minimum record type, then by record value, then by GID.
> -build SinglePos format 1, with a new subtable for each change of record type and value
> -build Single Pos format 2, with a new subtable for each change of record type
> -use whichever was smaller.

This doesn't sound like the best way to me.  The optimal is almost always a mix of the two subtable types.

**Comment by anthrotype** [comment]
> I think there's a bug open about otlLib.builder.buildSinglePos creating too many subtables..

this may be relevant in this regard https://github.com/fonttools/fonttools/pull/1539

**Comment by anthrotype** [comment]
Fixed by #1641
